python/lsst/daf/butler/_butler.py

```diff
@@ -1433,6 +1433,10 @@ def ingest(self, *datasets: FileDataset, transfer: Optional[str] = "auto", run:
             # associated with `dataset`.
             resolvedRefs: List[DatasetRef] = []
             for ref in dataset.refs:
+                if ref.dataId in groupedData[ref.datasetType]:
+                    raise RuntimeError(f"Ingest conflict. Dataset {dataset.path} has same DataId as "
+                                       f"other ingest dataset"
+                                       f" {groupedData[ref.datasetType][ref.dataId][0].path} ({ref.dataId})")
```

Comment:
`daf.butler.registry.ConflictingDefinitionError` might be a better exception type to raise here.